addons/survey/controllers/main.py

```diff
@@ -29,18 +29,24 @@ def _fetch_from_access_token(self, survey_token, answer_token):
         """ Check that given token matches an answer from the given survey_id.
         Returns a sudo-ed browse record of survey in order to avoid access rights
         issues now that access is granted through token. """
-        survey_sudo = request.env['survey.survey'].with_context(active_test=False).sudo().search([('access_token', '=', survey_token)])
-        if not answer_token:
-            answer_sudo = request.env['survey.user_input'].sudo()
-        else:
-            answer_sudo = request.env['survey.user_input'].sudo().search([
-                ('survey_id', '=', survey_sudo.id),
-                ('access_token', '=', answer_token)
-            ], limit=1)
-        return survey_sudo, answer_sudo
-
-    def _check_validity(self, survey_token, answer_token, ensure_token=True, check_partner=True):
-        """ Check survey is open and can be taken. This does not checks for
+        SurveySudo, UserInputSudo = request.env['survey.survey'].sudo(), request.env['survey.user_input'].sudo()
+        if not survey_token:
+            return SurveySudo, UserInputSudo
+        if answer_token:
+            answer_sudo = UserInputSudo.search(expression.AND([
+                [('survey_id', 'any', expression.AND([
+                    [('access_token', '=', survey_token)],
+                    [('active', 'in', (True, False))]  # keeping active test for UserInput
+                ]))],
+                [('access_token', '=', answer_token)],
+            ]), limit=1)
+            if answer_sudo:
+                return answer_sudo.survey_id, answer_sudo
+
+        return SurveySudo.with_context(active_test=False).search([('access_token', '=', survey_token)]), UserInputSudo
+
+    def _check_validity(self, survey_sudo, answer_sudo, answer_token, ensure_token=True, check_partner=True):
```

Comment:
I'm not a big fan of changing the method signature in stable... Couldn't it be done using a context key? :thinking:

Reply:
The target was changed to master ;-)

Reply:
Indeed! My bad, thanks :)